### PR TITLE
improve zoom logic

### PR DESCRIPTION
**Pull request by abrin**
we’re finding that the zoom in and zoom out logic isn’t working as expected, and that the buttons aren’t able to be initialized properly.  Creating a story to try and illustrate this, and add some explicit properties on the zoom to report whether we can zoom in or out manually.

I also added some explicit properties to the zoom event and the scaleInformation to make it clear how to set buttons

### CODE REVIEW

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ import { globalVault } from '@iiif/vault';
 
 export function useGenericAtlasProps<T = Record<never, never>>(props: GenericAtlasComponent<T>) {
   const webComponent = useRef<HTMLElement>();
+  const ZOOM_OUT_FACTOR = 0.75;
+  const ZOOM_IN_FACTOR = 1.0 / 0.75;
```
Comment (abrin) [inline packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:24]: These should be reversed in value... but the math doesn't make sense to me 🤯

File: packages/storybook/src/stories/canvas-panel.stories.tsx [hunk 2]
```diff
@@ -24,3 +24,38 @@ export const CanvasWithNavigator = () => {
...
+    const scaleInfo = panel?.getScaleInformation();
+    setCanZoomIn(scaleInfo.canZoomIn);
+    setCanZoomOut(scaleInfo.canZoomOut);
+    panel?.addEventListener("zoom", (e) => {
+      const detail = (e as any).detail;
+      // console.log(zoomInfo);
+      setZoomInfo(detail);
+      setCanZoomIn(detail.canZoomIn == false);
+      setCanZoomOut(detail.canZoomOut == false);
+
```
Comment (abrin) [inline packages/storybook/src/stories/canvas-panel.stories.tsx:48]: this logic seems to flash disabled then go to the current state, I have no idea why, it's frustrating

File: packages/storybook/src/stories/canvas-panel.stories.tsx [hunk 3]
```diff
@@ -26,12 +26,50 @@ export const CanvasWithNavigator = () => {
...
+    }
+  
+  useEffect(() => {
+    const panel = document.querySelector("canvas-panel");
+    // set the initial state based on the image that's loaded into the canvas
+    setTimeout(() => {
+      const scaleInfo = (panel as any).getScaleInformation();
+      setCanZoomIn(scaleInfo.canZoomIn);
+      setCanZoomOut(scaleInfo.canZoomOut);
+    },100)
```
Comment (abrin) [inline packages/storybook/src/stories/canvas-panel.stories.tsx:50]: @stephenwf  a question, is there a way in Canvas and Sequence Panel to know that it's loaded the manifest and atlas has rendered the world such that the scale info would be accurate?

Comment (stephenwf) [reply packages/storybook/src/stories/canvas-panel.stories.tsx:50]: This is something I've also found myself needing. I think Atlas needs to send more events generally. There are the "layout" events, such as `recalculate-world-size` which will fire when the world-size changes. But there is no `ready` or `images-loaded` events.

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts (hunk 1 above)
Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:24]: Yes, that's confusing. Because the viewport is like a box being drawn on the image - when you zoom out, the box gets bigger. When you zoom out, the box gets smaller. I get these mixed up myself often.

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:24]: ok, I feel less crazy now, thansk

File: packages/storybook/src/stories/canvas-panel.stories.tsx (hunk 3 above)
Comment (abrin) [reply packages/storybook/src/stories/canvas-panel.stories.tsx:50]: when I was logging the scaleInfo, I was seeing that the `_lastGoodScale` was null before the world had fully loaded, I think at least as a start we could set `isReady` in state when it is not null and emit an event.

**Comment by abrin** [comment]
I'll work with my team to test this a bit, but I _think_ if this works, this leaves our open issues to just the rotation issue Dan was hitting with the Bayard album images embedded in the canvas!

File: packages/storybook/src/stories/canvas-panel.stories.tsx (hunk 3 above)
Comment (abrin) [reply packages/storybook/src/stories/canvas-panel.stories.tsx]: turns out, that's already 'ready' by the time that CP gets access... found a few things, and exposed the repaint event with some of the same info which helps

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts [hunk 4]
```diff
@@ -176,24 +186,41 @@ export function useGenericAtlasProps<T = Record<never, never>>(props: GenericAtl
...
-                  min: minZoom,
                   minZoomCount,
+                  ...calculateZoomInformation(rt),
                   isMin,
                   isMax: Math.abs(rt.maxScaleFactor - rt._lastGoodScale) < 0.0002,
                   ...((data as any) || {}),
                 },
               })
             );
+            return;
```
Comment (abrin) [inline packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:197]: adding a return to escape things

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts [hunk 5]
```diff
@@ -176,24 +186,42 @@ export function useGenericAtlasProps<T = Record<never, never>>(props: GenericAtl
...
         }
       });
     }
 
     return () => void 0;
   }, [isReady]);
 
+  function calculateZoomInformation(rt?: Runtime) {
+    if (rt) {
+      console.log('zoom');
```
Comment (abrin) [inline packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: @stephenwf  -- This is an interesting issue... if you switch to a second canvas we're not seeing events fired anymore... weird no? I added a story (CanvasWithSkipSizes) to better illustrate this ... also seeing that the sizing issue might not be working for us.

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: it looks like when you change canvases the runtime gets wiped away and replaced, but nothing's bound anymore, so no more events

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: That's a really frustrating bug.
https://github.com/digirati-co-uk/iiif-canvas-panel/blob/9fb045e2e8f870cef47ac7c633252ac69370b2f0/packages/canvas-panel/src/hooks/use-generic-atlas-props.ts#L643-L645

We could add a counter to this, which will change when the RT is updated. That can be added to the second argument of `useEffect()` to rebind the events.

```ts
const [runtimeVersion, setRuntimeVersion] = useState(0);

// Inside the linked code when the runtime is set? Maybe with a:
// if (rt.runtime !== runtime.current) { ... }
// Check before.
setRuntimeVersion(n => n + 1);


useEffect(() => {
  ... 
}, [isReady, runtimeVersion])
```

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: Sequence panel _can_ persist the atlas instance (runtime etc) but I found that other bugs came up, and resetting everything was the smoothest option.

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: In the RIV, there is a prop to control if its reused, but we don't have that in canvas panel I don't think.

https://github.com/digirati-co-uk/react-iiif-vault/blob/69604aa044874e19f79f039213f4598cec8d677e/src/canvas-panel/index.tsx#L73

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:209]: let me dig into this a bit more... thanks!

File: packages/canvas-panel/src/hooks/use-generic-atlas-props.ts [hunk 6]
```diff
@@ -202,11 +203,10 @@ export function useGenericAtlasProps<T = Record<never, never>>(props: GenericAtl
     }
 
     return () => void 0;
-  }, [isReady]);
+  }, [isReady, runtimeVersion]);
```
Comment (abrin) [inline packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:206]: @stephenwf Ok.  so, that _kinda_ worked.  Now we're getting events from each canvas. BUT... we get n more events for every canvas we increment to... any ideas on this one?  I'm sorry, to bother you again on this

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:206]: Hmm, that could be an Atlas bug. The `rt.world.subscriptions` holds the event handlers, if this is growing too - it could be a problem there. I will try and pull down the changes and have a closer look too

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:206]: I tried adding a console log in CanvasPanel, and it looks like there are consistently 5 subscribers, but not 100% sure  which runtime that's one

Comment (stephenwf) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:206]: Maybe there is a missing clean up in Atlas, when the instance is replaced.

Comment (abrin) [reply packages/canvas-panel/src/hooks/use-generic-atlas-props.ts:207]: I found it!  it wasn't in atlas, or the canvas panel _(I added a queue in the event dispatch logic to dedup things, and couldn't find it)_ instead, it was in the story binding the useEffect to `[document.querySelector("canvas-panel")])` meant it was firing more often than it should, changing it to test for null or not removed them :(.  Clearly, I have a lot to learn about react

**Comment by abrin** [comment]
@stephenwf , ok I _think_ this is about ready -- pending any comments or recommendations you might have.  I've added a 'worldReady' event that fires when `_lastGoodScale` is set to something that's <1 and not undefined.  I've also cleaned up the story to try and make it a bit more useful when debugging (adding actions and more controls).  I've also double-checked the events firing are unique. 🤞 this should get the zoom and readiness logic working so we can better bind events, my basic testing in the story seems to show so.

**Comment by abrin** [comment]
still not seeing the `worldReady` working 100% of the time, sometimes in storybook it's logging out `detail: null` in the event, and sometimes in our app when using this build it's firing with a scaleFactor of 1 by polling until we don't see it... but we can work around it anyone doesn't have ideas.

**Comment by abrin** [comment]
@stephenwf  I've fixed a number of other small issues:
- I updated the content state loading for CanvasPanel to use a 'goto' to reflect that the 'target' approach broke changing canvases
- I found some cases with small images where the logic for zooming out wasn't working... I also started using a 'look-ahead' for the zoom out to see if we can better guess if zooming out will work or not.
- I adjusted the stories to be a bit more focused, and try and clean them up to make things easier to copy and adjust
- I added support for more events firing a zoom event to make sure that when a content-state is loaded, buttons can know to update
- I added guards around setContentStateText so we can set null or empty text without errors
- I'm still struggling with getting the logic for 'world-ready' to fire when the runtime logic is ready and get it seen in our app, I added a minor delay which seems to help ensure it only fires after the 'ready' fires... but also initially set it up on a way that doesn't really work in React. Instead I, binding to the first world resize event, which seems to be working

**Comment by stephenwf** [comment]
Thanks @abrin, I will have a look at this now. Are you happy for me to merge? The CodeSandbox is working again if you wanted to test the package. I can make a release too if required.

**Review by stephenwf — APPROVED** [review]

**Comment by abrin** [comment]
I'll run one last check and then let you know. I think we still have open questions on rotation and maybe one or two other things, but 🤞, we're getting close

**Comment by abrin** [comment]
we're good -- I'm seeing events firing, and can now bind other eventListeners to worldReady really logically.

**Comment by abrin** [comment]
@stephenwf if it's possible to merge this and give us an updated deploy!  I think also -- I probably owe you some updates to the documentation on the zoom event logic and the `world-ready` event

**Comment by stephenwf** [comment]
`@digirati/canvas-panel-web-components@1.0.61` published now 🎉  thanks again for these bug fixes and improvements!
